File: Python/147.Insertion Sort List.py
#用插入排序给一个链表排序
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def insertionSortList(self, head):
        if not head: return head
        nHead = ListNode(0)
        nHead.next , pre , cur = head , head , head.next
        while cur:
            post = cur.next
            prev , curr = nHead , nHead.next
            while curr!=cur:
                if curr.val >= cur.val:
                    break
                prev = curr
                curr = curr.next
            if curr!=cur:
                prev.next = cur
                cur.next = curr
                pre.next = post
            else:
                pre = cur
            cur = post
        return nHead.next

l1 = ListNode(3)
l2 = ListNode(2)
l3 = ListNode(1)
l4 = ListNode(9)
l5 = ListNode(6)
l6 = ListNode(11)
l1.next = l2
l2.next = l3
l3.next = l4
l4.next = l5
l5.next = l6

s = Solution()
p = s.insertionSortList(l1)
while p:
    print(p.val)
    p = p.next

File: C++/Linked List/147.Insertion Sort List.cpp
/*
用插入排序法给链表排序。
*/


#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


class Solution {
public:
	ListNode* insertionSortList(ListNode* head) {
        if (!head || !head->next)	return head;
		ListNode *nhead = new ListNode(0);
		nhead->next = head;
		ListNode *cur = head->next, *insert, *pre = head;
		while (cur)
		{
			insert = nhead;
			while (insert->next->val < cur->val && insert->next != cur)
			{
				insert = insert->next;
			}
			if (insert->next == cur)
			{
				pre = cur;
				cur = cur->next;
			}
			else
			{
				pre->next = cur->next;
				cur->next = insert->next;
				insert->next = cur;
				cur = pre->next;
			}
		}
		return nhead->next;
    }
};
int main()
{
	Solution s;
	ListNode head(3);
	ListNode t1(2);
	ListNode t2(8);
	ListNode t3(4);
	ListNode t4(10);
	ListNode t5(6);
	head.next = &t1;
	t1.next = &t2;
	t2.next = &t3;
	t3.next = &t4;
	t4.next = &t5;
 	ListNode* h = s.insertionSortList(&head);
 	while(h)
 	{
 		cout<<h->val<<endl;
 		h = h->next;
 	}
}

